File: scripts/ci/verify-deploy-manifest.mjs
#!/usr/bin/env node
/** Verify deploy-manifest.json fingerprint (used in CD before deploy). */
import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";

function stableStringify(value) {
  if (value === null || typeof value !== "object") return JSON.stringify(value);
  if (Array.isArray(value)) {
    return `[${value.map((item) => stableStringify(item)).join(",")}]`;
  }
  const keys = Object.keys(value).sort();
  return `{${keys.map((k) => `${JSON.stringify(k)}:${stableStringify(value[k])}`).join(",")}}`;
}

function fingerprintOf(payload) {
  const hash = createHash("sha256").update(stableStringify(payload)).digest("hex");
  return `sha256:${hash}`;
}

const path = process.argv[2] ?? "deploy-manifest.json";
const manifest = JSON.parse(readFileSync(path, "utf8"));
const { fingerprint, ...body } = manifest;

if (!fingerprint || typeof fingerprint !== "string") {
  console.error("manifest missing fingerprint");
  process.exit(1);
}

const expected = fingerprintOf(body);
if (fingerprint !== expected) {
  console.error(`fingerprint mismatch: got ${fingerprint}, expected ${expected}`);
  process.exit(1);
}

if (!body.gitSha || !body.artifacts?.api?.image || !body.artifacts?.api?.digest) {
  console.error("manifest missing required fields (gitSha, artifacts.api)");
  process.exit(1);
}

console.log(`manifest OK sha=${body.gitSha} image=${body.artifacts.api.image}`);

File: scripts/ci/verify_deploy_manifest.py
#!/usr/bin/env python3
"""Verify deploy-manifest.json fingerprint (CI/CD + VPS deploy; stdlib only)."""
from __future__ import annotations

import hashlib
import json
import sys
from pathlib import Path


def stable_dumps(payload: object) -> str:
    return json.dumps(payload, ensure_ascii=False, sort_keys=True, separators=(",", ":"))


def fingerprint_of(payload: object) -> str:
    digest = hashlib.sha256(stable_dumps(payload).encode("utf-8")).hexdigest()
    return f"sha256:{digest}"


def verify_manifest(path: Path) -> dict:
    manifest = json.loads(path.read_text(encoding="utf-8"))
    fingerprint = manifest.pop("fingerprint", None)
    if not isinstance(fingerprint, str):
        raise SystemExit("manifest missing fingerprint")

    expected = fingerprint_of(manifest)
    if fingerprint != expected:
        raise SystemExit(f"fingerprint mismatch: got {fingerprint}, expected {expected}")

    if not (
        manifest.get("gitSha")
        and manifest.get("artifacts", {}).get("api", {}).get("image")
        and manifest.get("artifacts", {}).get("api", {}).get("digest")
    ):
        raise SystemExit("manifest missing required fields (gitSha, artifacts.api)")

    return manifest


def main(argv: list[str]) -> int:
    path = Path(argv[1] if len(argv) > 1 else "deploy-manifest.json")
    manifest = verify_manifest(path)
    api_image = manifest["artifacts"]["api"]["image"]
    print(f"manifest OK sha={manifest['gitSha']} image={api_image}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))
